Graph-Kiota software requirements specifications (#58)
microsoftgraph/msgraph-sdk-design: graph-kiota/graph-kiota.md

> ### Graph OpenAPI specification

Reviewer: ```suggestion
### Graph OpenAPI description
```
The specification is the document that describes OpenAPI syntax.  API descriptions are what are created to describe APIs using the OpenAPI specification.

- OpenAPI v1 specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml

> - OpenAPI beta specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/beta/openapi.yaml

Reviewer 2: ```suggestion
- [OpenAPI beta specification](https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/beta/openapi.yaml)
```

### Graph libraries using Kiota

#### Graph Service library:

The Graph Service libraries which contain the request builders will be generated using Kiota

Generation command:

> ```

Reviewer 2: ```suggestion
```Shell
```

kiota.exe -d https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml --language typescript -o path\msgraph-sdk-typescript\src -c GraphServiceClient -n MicrosoftGraph
```

> - GraphServiceClient generated.

Reviewer 2: Can I suggest adding a link to the parameters reference? https://github.com/microsoft/kiota/#parameters-reference

- Request builders generated

**Graph Core library**:

 - Graph core library takes a dependency on Kiota core.
 - Kiota core contains code for a generic HTTP API. That is:
    - code to call make the native HTTP client call to the API
    - any generic middleware.
 - Graph specific code stays in Graph core. Example: Graph Tasks.

#### Design

Graph Core

// Following is for typescript only
``` typescript
GraphCoreClient {
    httpClient:  KiotaHttpClient // importing the HTTPClient from Kiota core library
}
```

GraphServiceClient
```
class GraphServiceClient{
    httpCoreInstance : HttpCore // From Kiota

    UserRequestBuilder();
    MailRequestBuilder();
}

```

#### Graph core library specifications

This library contains the
- Factory method/classes returning the:
    - Core client with the user configurations such as authprovider, custom middleware, request options.
    - HttpCore instance for the service library with user configurations such as authprovider, custom middleware, request options.

- Tasks
    - PageIteratorTask
    - LargeFileUploadTask
    - Batching

- Authentication Providers

>     - Use the `IAuthProvider` and  auth provider class from the Kiota Auth abstractions.

Reviewer 2: ```suggestion
    - Use the `IAuthenticationProvider` and  authentication provider class from the Kiota abstractions.
```

#### Graph Service library specifications

> - Discuss about ways to add custom code or custom configurations to the generated `GraphServiceClient` code.

Reviewer 2: I'm not sure I get what you mean by this statement, can you expand?

- The custom configurations that can go here are as follows: 
 - Setting the default for `HttpCore` with a Graph specific `HttpCore` instance. 
 - Adding multiple constructors to the `GraphServiceClient`. 

#### Graph Middlewares

- Import middlewares from Kiota core library

> - How to add graph specific checks in the middleware. Example: The JS SDK checks if the url is a Graph url or a custom host passed by the user before appending telemetry headers or auth headers. If not, the headers are removed.

Reviewer 2: the generic middleware could accept a set of valid hosts for example. @andrueastman did it for dotnet so the telemetry handler accepts a callback, effectively giving full control to graph core here. Some alignment work required across languages for sure.

> - AuthHandler - Kiota core

Reviewer 2: ```suggestion
- AuthHandler - removed
```
This is now handled by the auth providers:
- This avoids having to configure both the middleware and the provider
- The authentication provider implementations are available across any http client

- RetryHandler - Kiota core
- RedirectHandler - Kiota core
- CompressionHandler - Kiota core
- CachingHandler - Kiota core
- ChaosHandler - Graph core 
#### Use cases

- Graph core client 

> ```

Reviewer 2: ```suggestion
```TypeScript
```

(for all the snippets below)

var response = new coreClient.requestURL("/me").get();
```

- Graph service client
```

> var response = new serviceClient.me().get();

Reviewer 2: ```suggestion
var response = await serviceClient.me.get();
```

I believe we're generating getters for typescript for navigation like that. I'd have to double check

```

- Graph service client calling a core functionality

- The service library should extend the capabilities of the core library, that is, a user using service client should not require to create or use a core client to access any functionality. 
```
var response = new serviceClient.requestURL("/me").get(); // Example : a user of the JS SDK should be able to access GraphServiceClient.api(/me) without initialising the core.
```

- Graph core client with a task

```
var response = new coreClient.requestURL("/me").get();

PageIterator(coreclient, response)
```

- Graph service client with a task

```
var response = new serviceClient.me().get();
PageIterator(serviceClient, response)
```

- The service library should extend the capabilities of the core library, that is, a user using service client should not require to create or use a core client to access any functionality.
-  Some options to  achieve this (TBD):
 -  Maintain a parent abstract class for the core and the service client.
 -  `GraphServiceClient` extends the `GraphCoreClient`.

### More questions

> - Does the generated ApiClient have to be tightly coupled with the generated request builders? Should it always be tightly coupled? [MEM] I ask as there could be scenarios where the request builders are not necessary and the client developers just wants the Kiota.Core functionality in the client, and will use client.requestUrl<Response>.Get();

Reviewer 2: the api client is only a request builder with a special constructor and fewer properties. So in that sense yes. This was an early design choice.
